substratevm/DebugInfo.md

```diff
@@ -90,12 +165,346 @@ The currently implemented features include:
+$4 = {
+  <_arrhdrA> = {
+    hub = 0x906a78,
+    len = 2,
+    idHash = 0
+  }, 
+  members of java.lang.String[]:
+  data = 0x7ffff7c01038
+}
+```
+
+A simpler command which allows just the name of the hub object to be
+printed is as follows:
+
+```
+(gdb) x/s $2->hub->name->value->data
+798:	"[Ljava.lang.String;"
+```
+
+Indeed it is useful to define a gdb command `hubname` to execute this
+operation on an arbitrary input argument
+
+```
+command hubname
+  x/s (('java.lang.Class' *)((long)((('java.lang.Object' *)($arg0))->hub) & ~0x7L))->name->value->data
+end
+
+(gdb) hubname $2
+0x904798:	"[Ljava.lang.String;"
+```
+
+Notice that the `hubname` command also masks out the low 3 flag bits in
+the hub field that may sometimes get set by the runtime during program operation.
+
+Attempting to print the hub name for an invalid reference will fail
+safe, printing an error message.
+
+```
+(gdb) p/x (_objhdr *)$rdx
+$5 = 0x2
+(gdb) hubname $rdx
+Cannot access memory at address 0x2
+```
+
+Array type layouts are modelled with a class. The class inherits
+fields from an array header struct specific to the array element type,
+one of _arrhdrZ _arrhdrB, _arrhdrS, ... _arrhdrA (the last one is for
+object arrays). Inherited fields include the hub, array length, idHash
+to round up the header size to a boundary suitable for the array
+element type. The array class (layout) type includes only one field, a
+C++ array of length zero whose element type is a primtiive type or
+Java referece type.
+
+```
+(gdb) ptype 'java.lang.String[]'
+type = struct java.lang.String[] : public _arrhdrA {
+    java.lang.String *data[0];
+}
+```
+
+Notice that the type of the values stored in the data array is
+`java.lang.String *` i.e. the C++ array stores Java references
+i.e. addresss as far as the C++ model is concerned.
+
+The array header structs are all extensions of the basic _objhdr type
+which means that arrays and objects can both be safely cast to oops.
+
+```
+(gdb) ptype _arrhdrA
+type = struct _arrhdrA {
+    java.lang.Class *hub;
+    int len;
+    int idHash;
+}
+```
+
+Interfaces layouts are modelled as union types whose members are the
+layouts for all the classes which implement the interfacse.
+
+```
+(gdb) ptype 'java.lang.CharSequence'
+type = union _java.lang.CharSequence {
+    _java.lang.AbstractStringBuilder _java.lang.AbstractStringBuilder;
+    _java.lang.StringBuffer _java.lang.StringBuffer;
+    _java.lang.StringBuilder _java.lang.StringBuilder;
+    _java.lang.String _java.lang.String;
+    _java.nio.CharBuffer _java.nio.CharBuffer;
+}
+```
+
+Given a reference typed to an interface it can be resolved to the
+relevant class type by viewing it through the relevant union element.
+
+If we take the first String in the args array we can ask gdb to cast
+it to interface CharSequence
+```
+(gdb) print (('java.lang.String[]' *)$rdi)->data[0]
+$6 = (java.lang.String) 0x7ffff7c01060
+(gdb) print ('java.lang.CharSequence')$6
```

Comment:
Wow. This is nice.

Reply:
Yes, isn't it. I vaguely waved at this idea in my Linux Plumbers Conference talk when I was sketching out a Java -> C type mapping a a prior step to translating to DWARF. It's actually quite a neat way of looking at interfaces -- but only in a closed world!

I was really impressed  because someone in the audience immediately added that rider before I could :-)